content/api/cypress-api/custom-queries.md

```diff
@@ -0,0 +1,385 @@
+
+The name of the query you're adding.
+
+**<Icon name="angle-right"></Icon> callbackFn** **_(Function)_**
+
+Pass a function that receives the arguments passed to the query.
+
+This outer function is invoked once. It should return a function that takes a
+subject and returns a new subject; this inner function might be called multiple
+times.
+
+## Examples
+
+### `.focused()`
+
+The callback function can be thought of as two separate parts. The _outer
+function_, which is invoked once, where you perform setup and state management,
+and the _query function_, which might be called repeatedly.
+
+Let's look at an example. This is actual Cypress code - how `.focused()` is
+implemented internally. The only thing omitted here for simplicity is the
+TypeScript definitions.
+
+```javascript
+Commands.addQuery('focused', function focused(options = {}) {
+  const log = options.log !== false && Cypress.log({ timeout: options.timeout })
+
+  this.set('timeout', options.timeout)
+
+  return () => {
+    let $el = cy.getFocused()
+
+    log &&
+      cy.state('current') === this &&
+      log.set({
+        $el,
+        consoleProps: () => {
+          return {
+            Yielded: $el?.length ? $dom.getElements($el) : '--nothing--',
+            Elements: $el != null ? $el.length : 0,
+          }
+        },
+      })
+
+    if (!$el) {
+      $el = $dom.wrap(null)
+      $el.selector = 'focused'
+    }
+
+    return $el
+  }
+})
+```
+
+#### The outer function
+
+The outer function is called once each time test uses the query. It performs
+setup and state management:
+
+```javascript
+function focused(options = {}) {
+  const log = options.log !== false && Cypress.log({ timeout: options.timeout })
+
+  this.set('timeout', options.timeout)
+
+  return () => {} // Inner function
+}
+```
+
+Piece by piece, then.
+
+```javascript
+function focused(options = {}) {}
+```
+
+Cypress passes the outer function whatever arguments the user calls it with; no
+processing or validation is done on the user's arguments. In our case,
+`.focused()` accepts one optional argument, `options`.
+
+If you wanted to validate the incoming arguments, you might write something
+like:
+
+```javascript
+if (options === null || !_.isPlainObject(options)) {
+  const err = `cy.root() requires an \`options\` object. You passed in: \`{options}\``
+  throw new TypeError(err)
+}
+```
+
+This is a general pattern: when something goes wrong, queries just throw an
+error. Cypress will handle displaying this in the Command Log nicely.
+
+```javascript
+const log = options.log !== false && Cypress.log({ timeout: options.timeout })
+```
+
+If the user has not set `{ log: false }`, we create a new `Cypress.log()`
+instance. See [`Cypress.log()`](/api/cypress-api/cypress-log) for more
+information.
+
+This is setup code - we only want it to run once, creating the log message when
+Cypress first begins executing this query. We hold onto a reference to `Log`
```

Comment:
Does this mean these log references always exist until a test has completed due to the command chain re-executing the inner function?

Reply:
I'm not entirely sure what you're getting at? Yes, the closure holds onto a reference to the $Log object, so as long as the closure is referenced, $Log can't be GCed.

Existing commands do the exact same thing, though they mostly store their reference as $Command.attributes._log, rather than in a closure. Changing the place we hold onto a reference (as a closure local variable rather than as an attribute on an object) shouldn't have any memory-usage implications.